File: Backend/backend/Models/Loan.cs
namespace Backend.Models;

using Backend.DTOs;
using System.ComponentModel.DataAnnotations.Schema;
using System.Text.Json.Serialization;

public class Loan : BaseModel
{
    [JsonIgnore]
    public int CopyId { get; set; }
    public Copy Copy { get; set; } = null!;
    [JsonIgnore]
    public int UserId { get; set; }
    [JsonIgnore]
    public User User { get; set; } = null!;
    public UserResponseDTO? UserInfo => User != null ? UserResponseDTO.FromUser(User) : null;
    public DateTime LoanedAt { get; set; }
    public DateTime DueDate { get; set; }
    public bool Returned { get; set; } = false;
}

File: frontend/src/types/loan.ts
import { Base } from "./base";
import { Copy } from "./copy";
import { Pagination } from "./pagination";
import { User } from "./user";

export interface Loan extends Base {
    userInfo:User,
    copy:Copy,
    loanedAt:Date,
    dueDate:Date,
    returned:boolean
}

export type LoanFilter = {
    filter: "OnGoing" | "Expired" | null,
    pagination: Pagination | null,
    userId: number |null
}

export type MakeLoan = {
    userId: number,
    copyId: number
}

export type UpdateLoan = {
    id: number,
    userId: number,
    returned: boolean | null,
    dueDate: Date | null
}